Cython/Compiler/ExprNodes.py

```diff
@@ -1997,6 +1997,11 @@ def declare_from_annotation(self, env, as_target=False):
             atype = unspecified_type if as_target and env.directives['infer_types'] != False else py_object_type
         if atype.is_fused and env.fused_to_specific:
             atype = atype.specialize(env.fused_to_specific)
+        if as_target and env.is_c_class_scope and not atype.is_pyobject:
+            # TODO: this will need revising slightly for cdef dataclasses when implemented
```

Comment:
And probably also if we decide to allow the annotation syntax for defining cdef attributes. That's not entirely out of focus.